### PR TITLE
ball vision python

### PR DESCRIPTION
**Pull request by henrymwestfall**

### CODE REVIEW

**Review by michael-brennan2005 — APPROVED** [review]

File: vision/high_goal_python/high_goal.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    :param targets: a list of targets, each represented with a tuple containing its contour and bounding rect
+
+    :return: a rectangle around the target containing the highest point (or None if no targets are passed)
+    """
+    absolute_highest = None
+    highest_target = None
+    for target_contour, target_rect in targets:
+        if target_rect.center_y > center_y:  # targets must be above middle of image or else they are false positives
+            highest_in_contour = max(target_contour, key=lambda point: point[1])
+            if absolute_highest is None or highest_in_contour[1] > absolute_highest[1]:
```
Comment (Anonymoustank) [inline vision/high_goal_python/high_goal.py:115]: comparing y-values with 1st index

File: vision/high_goal_python/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+def apply_hsv_filter(config: dict, image: np.ndarray) -> np.ndarray:
+    """
+    Get a bit mask from the image containing points that pass an HSV filter.
+
+    :param config: the configuration parameters for the filtering
+    :param image: the image to filter
+    :return: a new image containing the bit mask
+    """
+    # convert image to HSV
+    image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
```
Comment (Anonymoustank) [inline vision/high_goal_python/pipeline.py:39]: COLOR_BGR2HSV = color conversion code to switch between RGB and HSV

File: vision/high_goal_python/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import cv2
+import numpy as np
+from math import tan, atan, hypot
+from typing import List
+
+# kernel for morphological transformations
```
Comment (Anonymoustank) [inline vision/high_goal_python/pipeline.py:6]: higher kernel = more blur

File: vision/high_goal_python/rectangle.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        self.box_points = box_points
+
+        # calculate center point
+        # use built-in sum instead of numpy sum because numpy will add individual point data into a single *value*
+        # while build-in sum will add together points as vectors to get a single *point* as the result
+        self.center = sum(self.box_points) / 4  # boxes have 4 corners
+
+        # calculate rectangle height, width, area, and rotation angle
+        side = self.box_points[1:3]  # second and third points form a vertical side
+        self.height = np.linalg.norm(side)
```
Comment (Anonymoustank) [inline vision/high_goal_python/rectangle.py:21]: distance of vector (side length)

**Pull request merged**